### PR TITLE
Refactoring + World Gen + Lighting

### PR DESCRIPTION
**Pull request by Col-E**
### Whats new:

* WIP biome generation with an [example implementation in terrain tasks](https://files.gitter.im/last-try/Lobby/MMKi/image.png).
* Simplex-noise based [cave generation](https://my.mixtape.moe/uixgtz.mp4)
   * Bias for [surface generation](https://my.mixtape.moe/elwqgz.mp4) keeps the surface clean and smooth
* WIP lighting system.
   * Works based off of a simple neighboring block & wall calculation 
   * Updates on world start, and only does small edits when blocks and such are placed
   * Doesn't support day/night cycles 
   * Doesn't support light sources like torches
* Launch arg `-wd` - Deletes world on startup so you can easily test world-gen
* Launch arg `-f` for fullscreen
* Launch arg `-nl` to disable the WIP lighting
* You can edit the window size in `DesktopLauncher`

### Whats changed:

* Encapsulated `Globals.world` for later possibilities of swapping between worlds
* Encapsulated world components for hooks to be inserted later *(Like for mods & light updates)*
* Inventory is now easy to implement. Simply inherit the Inventory interface for inventories *(Player inventory, chests, etc.)*, or InventoryOwner for entities that contain inventories *(Like merchants and such)*
    * Basically all of the code has been modularized and put into `org.egordorichev.lasttry.inventory`
* World generation now is based off of seed values. This allows for recreating the same world multiple times.
* Chat handling has been rewritten.
   * Future plan: Hide certain commands from the player if debug mode not enabled
* Opening / closing screens has been centralized `Globals.setCurrentScreen(screen)`

### CODE REVIEW

**Comment by egordorichev** [comment]
Hey, @Col-E! Big thank you, that you are here, while i'm offline. As @Nufllee sad, I will be back on night from 4 to 5 june. Right now, i'm trying to join gittr from my book reader ;)

I really like the -wd feature.
Also, can you add generation of surface biomes to the algorithm?

Big thanks, Egor.

**Comment by egordorichev** [comment]
Code climat is broken right now, you should check out code beat! I'm amazed, with new inventory system and lighting! I can't merge pull right now, but you should be able to do it your self.

**Comment by Col-E** [comment]
No problem!

> can you add generation of surface biomes to the algorithm

Currently I have a rough biome GeneratorTask layed out but it doesn't focus on the surface. Its based on raw noise so biomes can be put anywhere. I'm not exactly sure how I would force it to populate the surface like terraria. I'd have to look at their code more.

>  I can't merge pull right now, but you should be able to do it your self.

I'm gonna hold off until I can fix / add a few more things before I merge. Hope you're enjoying your vacation.

**Comment by egordorichev** [comment]
Also, why did you add temperature/humidy to biomes? We handle them by calculsting on-screen blocks.
Also, you've made a todo, for storing them in json. That can be easily done, by changingv hardcoded values in Envirnoment class.

Here is how you can store them:

```
{ 
  "Corruption" : {
    "background" : "…",
    "required" : {
       { "id" : 24, "count" : 200 }
    }
}
```

This will make more sense...

**Comment by Col-E** [comment]
The biome temp/humid is used for initial world generation. You need to have an algorithm for placing the biomes initially. The temps and humidites are not used anywhere outside of generation.

https://www.gamedev.net/topic/682351-biome-based-map-generation/

Basically it ensures biomes like tundras and jungles wont be placed next to each other.

**Comment by Col-E** [comment]
I guess we can have both. When the player creates a world just have the option for *"Classic terraria biome placement"* and *"Natural biome placement"*. Never hurts to provide extra.
